pkg/optimizelyjson/optimizely_json.go

```diff
@@ -0,0 +1,98 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package optimizelyjson //
+package optimizelyjson
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// OptimizelyJSON holds the underlying structure of the object
+type OptimizelyJSON struct {
+	jsonRepr map[string]interface{}
+}
+
+// NewOptimizelyJSON constructs the object
+func NewOptimizelyJSON(jsonRepr map[string]interface{}) *OptimizelyJSON {
+	return &OptimizelyJSON{jsonRepr: jsonRepr}
+}
```

Comment:
We haven't tied in the project config yet, but I think this should probably be instantiated with `payload string` since that is what we are storing in the datafile. Then, depending on the method calls below we'd, memoize the transformations into `map[string]interface{}` if that was ever requested. As it's written now I feel we'd be marshaling data back and forth unnecessarily.

Reply:
> We haven't tied in the project config yet, but I think this should probably be instantiated with `payload string` since that is what we are storing in the datafile. Then, depending on the method calls below we'd, memoize the transformations into `map[string]interface{}` if that was ever requested. As it's written now I feel we'd be marshaling data back and forth unnecessarily.

I thought about that too, but then I thought that 2 out of 3 methods need map[string]interface{}, and only one method needs it as a string. But I agree that public SDK's would need to take the string and Unmarshal into that interface.

Reply:
I have implemented string payload, with map[string]interface{} used for memoization. I did look at the reflect package and wanted to avoid this memoization. Things get really complicated when the user does not provide a json tag. Unmarshal() can still match by field name, case insensitive. That would require creating a map of field names or something like that (steal some logic from Unmarshal()...), which I don't think will be that pretty if you want to preserve that functionality.  Also, I don't think we can enforce the user to provide json tags all the time.